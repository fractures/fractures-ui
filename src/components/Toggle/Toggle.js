import cc from "classcat"
import Label from "../Label/Label"
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"

const propTypes = {
	check: PropTypes.func.isRequired,
	className: PropTypes.string,
	isChecked: PropTypes.bool,
	isDisabled: PropTypes.bool,
	label: PropTypes.string.isRequired
}

const defaultProps = {
	check: () => null,
	isChecked: false,
	isDisabled: false,
	label: ""
}

const NakedToggle = props => {
	const checkboxClasses = cc({
		"fr-toggle": true,
		"fr-toggle--checked": props.isChecked,
		"fr-toggle--unchecked": !props.isChecked,
		"fr-toggle--disabled": props.isDisabled
	})

	return (
		<div className={ props.className }>
			<label className={ checkboxClasses }>
				<input type="checkbox" defaultChecked={ props.isChecked } onChange={ () => props.check() } />
				<span className="fr-toggle__mark" />
				<Label>{props.label}</Label>
			</label>
		</div>
	)
}

const Toggle = styled(NakedToggle)`
	box-sizing: border-box;

	user-select: none;

	label {
		cursor: pointer;
	}

	input {
		display: none;
		height: 0;
		width: 0;

		visibility: none;
	}

	.fr-toggle {
		position: relative;

		display: flex;
	}

	.fr-toggle--disabled {
		opacity: 0.5;
		pointer-events: none;
	}

	.fr-toggle__mark {
		box-sizing: border-box;
		height: 1.5rem;
		margin-right: 0.5rem;
		overflow: hidden;
		width: 3rem;

		background-color: var(--fr-500);
		border-radius: 3rem;
	}

	.fr-toggle__mark:after {
		left: 0.25rem;
		top: 0.25rem;
		position: absolute;

		box-sizing: inherit;
		height: 1rem;
		width: 1rem;

		background-color: white;
		border-radius: 100%;
		transition: all 100ms ease-in-out;
		content: " ";
	}

	.fr-toggle__mark:before {
		left: 0;
		top: 0;
		position: absolute;

		box-sizing: inherit;
		padding: 0 0.5rem;
		width: 3rem;

		color: white;

		content: "Off";
		font-size: 0.75rem;
		font-weight: 500;
		line-height: 1.5rem;
		text-align: right;
	}

	input:checked ~ .fr-toggle__mark:after {
		left: 1.75rem;
	}

	input:checked ~ .fr-toggle__mark:before {
		content: "On";
		text-align: left;
	}

	input:checked ~ .fr-toggle__mark {
		background-color: var(--fr-900);
	}
`

NakedToggle.propTypes = propTypes
NakedToggle.defaultProps = defaultProps

Toggle.propTypes = propTypes
Toggle.defaultProps = defaultProps

export default Toggle
