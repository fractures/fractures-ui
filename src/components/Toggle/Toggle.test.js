import Toggle from "./Toggle"

const mockFn = jest.fn()

describe("Toggle", () => {
	const componentRender = <Toggle label="Toggle" check={ mockFn() } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})
})
