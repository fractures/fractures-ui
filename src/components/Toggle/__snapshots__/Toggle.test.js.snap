// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Renders correctly 1`] = `
<div
  className="sc-bwzfXH keNWvK"
>
  <label
    className="hl-toggle hl-toggle--unchecked"
  >
    <input
      defaultChecked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="hl-toggle__mark"
    />
    <span
      className="sc-bdVaJa ehcbJe"
    >
      Toggle
    </span>
  </label>
</div>
`;
